### PR TITLE
[openvdb7] Add openvdb7 for openvdb 7.x

### PR DESCRIPTION
**Pull request by archibate**
**Describe the pull request**

- #### What does your PR fix?  
  Fixes nothing, this PR adds a new port of `openvdb` with version 7.2.3, given that [our software](https://github.com/zenustech/zeno) have known bug with openvdb 8.x (which seems to be recently upgraded in vcpkg).

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all, Yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  `Yes, I see.`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
Yes, thank for the reminder.

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by archibate** [comment]
Why is CI failing? Is it about building OpenVDB or some config files I forgot to change?

**Comment by JonLiu1993** [comment]
@archibate ，Thanks for your pr, we already have the port of openvdb, why do we need to add a new port separately for this 7.2.3 version?

**Comment by JackBoosY** [comment]
> @archibate ，Thanks for your pr, we already have the port of openvdb, why do we need to add a new port separately for this 7.2.3 version?

We accpet the old version of our ports to be a new port.

File: ports/openvdb7/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,66 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO zenustech/openvdb
+    REF 8f289924a081a3fd9f5fca1db4826944a8450fe3 # v7.2.3
+    HEAD_REF master
+    PATCHES
+        0003-fix-cmake.patch
+)
```
Comment (JackBoosY) [inline ports/openvdb7/portfile.cmake:8]:
```
CMake Error at scripts/cmake/vcpkg_from_github.cmake:86 (message):
  SHA512 must be specified if REF is specified.
Call Stack (most recent call first):
  ports/openvdb7/portfile.cmake:1 (vcpkg_from_github)
  scripts/ports.cmake:142 (include)
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by archibate** [comment]
Oh I see, but how can I get the SHA256 of openvdb7? README says I should set SHA256 to 0 first then package it... not very sure what does this mean. Maybe add some command line utilities (e.g. `./vcpkg calc-sha256 openvdb7`) to help users figure out?











无法顺畅的大口呼吸，是活着的最好证明

---Original---
From: ***@***.***&gt;
Date: Fri, Nov 19, 2021 17:16 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [microsoft/vcpkg] [openvdb7] Add openvdb7 for openvdb 7.x (PR #21536)




 
@JackBoosY requested changes on this pull request.
 
 
In ports/openvdb7/portfile.cmake:
 &gt; +vcpkg_from_github( +    OUT_SOURCE_PATH SOURCE_PATH +    REPO zenustech/openvdb +    REF 8f289924a081a3fd9f5fca1db4826944a8450fe3 # v7.2.3 +    HEAD_REF master +    PATCHES +        0003-fix-cmake.patch +)  CMake Error at scripts/cmake/vcpkg_from_github.cmake:86 (message):   SHA512 must be specified if REF is specified. Call Stack (most recent call first):   ports/openvdb7/portfile.cmake:1 (vcpkg_from_github)   scripts/ports.cmake:142 (include)  
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by JackBoosY** [comment]
> Oh I see, but how can I get the SHA256 of openvdb7? README says I should set SHA256 to 0 first then package it... not very sure what does this mean. Maybe add some command line utilities (e.g. `./vcpkg calc-sha256 openvdb7`) to help users figure out?


Download this file and use command `./vcpkg hash FILE_FULL_PATH`.

**Comment by archibate** [comment]
As I wrote in PR description, openvdb have an API break from 7.x to 8.x. Locking to a working version for dependecies is very important for application developers (e.g. one day a dependecy bumped version and all our CI fails). So it would be great if vcpkg could support specifying which version to install, otherwise we may have to make multiple ports for each major version. In fact even minor version bump may cause API break, for example the [OpenSim](https://github.com/opensim-org/opensim-core) project worked with spdlog v1.4.1, but complie error in latest v1.12.0.. So really great if vcpkg could support specifying versions (e.g. `openvdb&gt;=7.0<8.0`) for each port..











无法顺畅的大口呼吸，是活着的最好证明

---Original---
From: ***@***.***&gt;
Date: Fri, Nov 19, 2021 17:15 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [microsoft/vcpkg] [openvdb7] Add openvdb7 for openvdb 7.x (PR#21536)




  
@archibate ，Thanks for your pr, we already have the port of openvdb, why do we need to add a new port separately for this 7.2.3 version?
  
We accpet the old version of our ports to be a new port.
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by archibate** [comment]
Cool, but which file?











无法顺畅的大口呼吸，是活着的最好证明

---Original---
From: ***@***.***&gt;
Date: Fri, Nov 19, 2021 17:58 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [microsoft/vcpkg] [openvdb7] Add openvdb7 for openvdb 7.x (PR #21536)




  
Oh I see, but how can I get the SHA256 of openvdb7? README says I should set SHA256 to 0 first then package it... not very sure what does this mean. Maybe add some command line utilities (e.g. ./vcpkg calc-sha256 openvdb7) to help users figure out?
  
Download this file and use command ./vcpkg hash FILE_FULL_PATH.
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

File: ports/openvdb7/portfile.cmake (hunk 1 above)
Comment (JackBoosY) [inline ports/openvdb7/portfile.cmake:8]:
```suggestion
vcpkg_from_github(
    OUT_SOURCE_PATH SOURCE_PATH
    REPO zenustech/openvdb
    REF 8f289924a081a3fd9f5fca1db4826944a8450fe3 # v7.2.3
    SHA512 776558fdd911f0cc9e8d467bf8e00a1930d2e51bb8ccd5f36f95955fefecab65faf575a80fdaacfe83fd32808f8b9c2e0323b16823e0431300df7bc0c1dfde12
    HEAD_REF master
    PATCHES
        0003-fix-cmake.patch
)
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/openvdb7/portfile.cmake (hunk 1 above)
Comment (JackBoosY) [reply ports/openvdb7/portfile.cmake:8]: After apply this request, you will see the correct hash value in the pipeline test results.

File: ports/openvdb7/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,67 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO zenustech/openvdb
+    REF 8f289924a081a3fd9f5fca1db4826944a8450fe3 # v7.2.3
+    SHA512 776558fdd911f0cc9e8d467bf8e00a1930d2e51bb8ccd5f36f95955fefecab65faf575a80fdaacfe83fd32808f8b9c2e0323b16823e0431300df7bc0c1dfde12
```
Comment (JackBoosY) [inline ports/openvdb7/portfile.cmake:5]:
```suggestion
    SHA512 4d0d45e761c24aa7f5383e665f77aee6f66520675f6c13bee3fa86f86fdd87947f606b856507968787d9a09ed9c25759e4c083cb0ea2c6df4e53febb43290113 
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by JackBoosY** [comment]
```
2021-11-24T06:38:11.2139448Z Starting package 82/82: openvdb7:x64-linux
2021-11-24T06:38:11.2140075Z Building package openvdb7[core]:x64-linux...
2021-11-24T06:38:12.8111499Z -- Downloading https://github.com/zenustech/openvdb/archive/8f289924a081a3fd9f5fca1db4826944a8450fe3.tar.gz -> zenustech-openvdb-8f289924a081a3fd9f5fca1db4826944a8450fe3.tar.gz...
2021-11-24T06:38:12.8112656Z -- Extracting source /mnt/vcpkg-ci/downloads/zenustech-openvdb-8f289924a081a3fd9f5fca1db4826944a8450fe3.tar.gz
2021-11-24T06:38:12.8113403Z -- Applying patch 0003-fix-cmake.patch
2021-11-24T06:38:12.8113959Z CMake Error at scripts/cmake/z_vcpkg_apply_patches.cmake:61 (message):
2021-11-24T06:38:12.8114824Z   Applying patch failed: Checking patch CMakeLists.txt...
2021-11-24T06:38:12.8115103Z 
2021-11-24T06:38:12.8115747Z   Hunk #1 succeeded at 335 (offset -19 lines).
2021-11-24T06:38:12.8116043Z 
2021-11-24T06:38:12.8116418Z   Checking patch cmake/FindOpenVDB.cmake...
2021-11-24T06:38:12.8116715Z 
2021-11-24T06:38:12.8117148Z   error: while searching for:
2021-11-24T06:38:12.8117387Z 
2021-11-24T06:38:12.8117790Z   
2021-11-24T06:38:12.8117975Z 
2021-11-24T06:38:12.8118390Z   # Add standard dependencies
2021-11-24T06:38:12.8118610Z 
2021-11-24T06:38:12.8118969Z   
2021-11-24T06:38:12.8119138Z 
2021-11-24T06:38:12.8119543Z   find_package(TBB REQUIRED COMPONENTS tbb)
2021-11-24T06:38:12.8119789Z 
2021-11-24T06:38:12.8120132Z   
2021-11-24T06:38:12.8120301Z 
2021-11-24T06:38:12.8120742Z   if(NOT OPENVDB_USE_STATIC_LIBS AND NOT Boost_USE_STATIC_LIBS)
2021-11-24T06:38:12.8121087Z 
2021-11-24T06:38:12.8121526Z     # @note  Both of these must be set for Boost 1.70 (VFX2020) to link against
2021-11-24T06:38:12.8121813Z 
2021-11-24T06:38:12.8122169Z   
2021-11-24T06:38:12.8122338Z 
2021-11-24T06:38:12.8122777Z   error: patch failed: cmake/FindOpenVDB.cmake:457
2021-11-24T06:38:12.8123031Z 
2021-11-24T06:38:12.8123469Z   error: cmake/FindOpenVDB.cmake: patch does not apply
2021-11-24T06:38:12.8123727Z 
2021-11-24T06:38:12.8124139Z   Checking patch openvdb/openvdb/CMakeLists.txt...
2021-11-24T06:38:12.8124392Z 
2021-11-24T06:38:12.8124789Z   error: while searching for:
2021-11-24T06:38:12.8125018Z 
2021-11-24T06:38:12.8125384Z     endif()
2021-11-24T06:38:12.8125590Z 
2021-11-24T06:38:12.8125958Z   endif()
2021-11-24T06:38:12.8126138Z 
2021-11-24T06:38:12.8126526Z   
2021-11-24T06:38:12.8126722Z 
2021-11-24T06:38:12.8127176Z   find_package(TBB ${MINIMUM_TBB_VERSION} REQUIRED COMPONENTS tbb)
2021-11-24T06:38:12.8127996Z 
2021-11-24T06:38:12.8128453Z   if(OPENVDB_FUTURE_DEPRECATION AND FUTURE_MINIMUM_TBB_VERSION)
2021-11-24T06:38:12.8128865Z 
2021-11-24T06:38:12.8129312Z     if(${Tbb_VERSION} VERSION_LESS FUTURE_MINIMUM_TBB_VERSION)
2021-11-24T06:38:12.8129842Z       message(DEPRECATION "Support for TBB versions < ${FUTURE_MINIMUM_TBB_VERSION} "
2021-11-24T06:38:12.8130133Z 
2021-11-24T06:38:12.8130486Z   
2021-11-24T06:38:12.8130656Z 
2021-11-24T06:38:12.8131104Z   error: patch failed: openvdb/openvdb/CMakeLists.txt:120
2021-11-24T06:38:12.8131423Z 
2021-11-24T06:38:12.8131850Z   error: openvdb/openvdb/CMakeLists.txt: patch does not apply
2021-11-24T06:38:12.8132129Z 
2021-11-24T06:38:12.8132550Z   Checking patch openvdb/openvdb/cmd/CMakeLists.txt...
2021-11-24T06:38:12.8132808Z 
2021-11-24T06:38:12.8133215Z   Hunk #1 succeeded at 50 (offset 4 lines).
2021-11-24T06:38:12.8133445Z 
2021-11-24T06:38:12.8133857Z   Hunk #2 succeeded at 120 (offset 4 lines).
2021-11-24T06:38:12.8134086Z 
2021-11-24T06:38:12.8134502Z Call Stack (most recent call first):
2021-11-24T06:38:12.8134995Z   scripts/cmake/vcpkg_extract_source_archive.cmake:230 (z_vcpkg_apply_patches)
2021-11-24T06:38:12.8135575Z   scripts/cmake/vcpkg_extract_source_archive_ex.cmake:32 (vcpkg_extract_source_archive)
2021-11-24T06:38:12.8136170Z   scripts/cmake/vcpkg_from_github.cmake:182 (vcpkg_extract_source_archive_ex)
2021-11-24T06:38:12.8136729Z   ports/openvdb7/portfile.cmake:1 (vcpkg_from_github)
2021-11-24T06:38:12.8137180Z   scripts/ports.cmake:142 (include)
2021-11-24T06:38:12.8137398Z 
2021-11-24T06:38:12.8137609Z 
2021-11-24T06:38:12.8138302Z Error: Building package openvdb7:x64-linux failed with: BUILD_FAILED
2021-11-24T06:38:12.8138997Z Elapsed time for package openvdb7:x64-linux: 1.593 s
```
You should also update the patches.

File: ports/openvdb7/0003-fix-cmake.patch [hunk 3]
```diff
@@ -38,16 +38,16 @@ index 1dc5bc0..d84d8f1 100644
...
- if(OpenVDB_USES_IMATH_HALF)
--  find_package(IlmBase REQUIRED COMPONENTS Half)
+ if(OpenVDB_USES_EXR)
+-  find_package(IlmBase REQUIRED)
 +  find_package(IlmBase CONFIG REQUIRED)
-   if(WIN32)
-     # @note OPENVDB_OPENEXR_STATICLIB is old functionality and should be removed
-     if(OPENEXR_USE_STATIC_LIBS OR (${ILMBASE_LIB_TYPE} STREQUAL STATIC_LIBRARY))
-@@ -664,7 +664,7 @@ set(_OPENVDB_HIDDEN_DEPENDENCIES)
+   find_package(OpenEXR REQUIRED)
```
Comment (archibate) [inline ports/openvdb7/0003-fix-cmake.patch:47]: This is a new diff in VDB 7.2.3.

File: ports/openvdb7/vcpkg-cmake-wrapper.cmake.in [hunk 4]
```diff
@@ -0,0 +1,12 @@
+set(OPENVDB_PREV_MODULE_PATH ${CMAKE_MODULE_PATH})
```
Comment (JackBoosY) [inline ports/openvdb7/vcpkg-cmake-wrapper.cmake.in:1]: This setting is used to use the installed find module normally. I don't think we need this.

**Review by JackBoosY — CHANGES_REQUESTED** [review]
Please run command `vcpkg x-add-version --all` then commit the changes.

**Comment by JonLiu1993** [comment]
@archibate, Have you tried @JackBoosY 's review suggestions?

File: ports/openvdb7/vcpkg-cmake-wrapper.cmake.in (hunk 4 above)
Comment (JackBoosY) [reply ports/openvdb7/vcpkg-cmake-wrapper.cmake.in:1]: @archibate I think we need to remove this cmake wrapper.

Comment (JonLiu1993) [reply ports/openvdb7/vcpkg-cmake-wrapper.cmake.in:1]: @archibate ,Could you reply to @JackBoosY 's review suggestion?

**Comment by JackBoosY** [comment]
Ping for response.

**Comment by JackBoosY** [comment]
Draft this PR since there is no response.

**Comment by JonLiu1993** [comment]
@archibate , any updates?

**Comment by archibate** [comment]
hi, sorry for forgeting this thread. I recently found the manifest mode of vcpkg (using vcpkg.json) when playing Zhihu, it supports specifying version<= and so on. So it seems like adding another port openvdb7 is reluctant, am I correct? Is manifest mode very mature now?











无法顺畅的大口呼吸，是活着的最好证明

---Original---
From: ***@***.***&gt;
Date: Wed, Feb 23, 2022 14:09 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [microsoft/vcpkg] [openvdb7] Add openvdb7 for openvdb 7.x (PR#21536)




 
@archibate , any updates?
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by JackBoosY** [comment]
```
it supports specifying version<= and so on
```
Only support `>=` now. See https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md#version-field
Manifest still need to improve more.

**Comment by JackBoosY** [comment]
Closing this PR since it seems that no progress is being made. Please reopen if work is still being done.

**Pull request closed without merging**